lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+
+      // If we have already seen this block...
+      if (!seenBlocks.insert(block).second)
+        continue; // skip it!
+
+      // Determine if this block is initializing, that is:
+      //
+      //     merge(OutAvailability(block.pred1, block.pred2, ...)) != Yes
+      //               AND
+      //     OutAvailability(block)) = Yes
+
+      auto outSet = getBlockInfo(block).OutAvailability;
+      if (!outSet.isAllYes())
+        continue; // then this block never sees TheMemory fully-initialized.
+
+      AvailabilitySet inSet(outSet.size());
+      for (const auto &pred : block->getPredecessorBlocks())
+        inSet.mergeIn(getBlockInfo(pred).OutAvailability);
+
+      if (inSet.isAllYes())
+        continue; // then this block always receives TheMemory fully-initialized.
+
+      initializingBlocks.insert({block, inSet});
+    } // end loop
+
+#ifndef NDEBUG
+    // dump the data-structure we just built
+    for (auto &entry : initializingBlocks) {
+      auto &block = entry.getFirst();
+      llvm::dbgs() << "in " << block->getParent()->getName()
+                   << ", bb" << block->getDebugID()
+                   << " is an initializing block with in-availability:\t";
+      entry.getSecond().dump(llvm::dbgs());
+      llvm::dbgs() << "\n";
+    }
+#endif
+  }
+  ////// end of Step 1
+
+  // Step 2: Scan each block to find the first non-load use that
+  // fully-initializes TheMemory, using the Map from Step 1.
+
+  // Tracks status of each element of TheMemory as we scan through the block.
+  AvailabilitySet localAvail;
+
+  for (auto const& entry : initializingBlocks) {
+    auto *block = entry.getFirst();
+
+    // set-up initial availability at the block's entry-point.
+    localAvail = entry.getSecond();
+
+    auto BBI = block->begin(); // our cursor and eventual insertion-point.
+    for (auto E = block->end(); BBI != E; ++BBI) {
+      auto *inst = &*BBI;
+
+      auto result = NonLoadUses.find(inst);
+      if (result == NonLoadUses.end())
+        continue; // not a possible store
+
+      // Mark the tuple elements involved in this use as defined.
+      for (unsigned use : result->second) {
+        auto const& instUse = Uses[use];
+        for (unsigned i = instUse.FirstElement; i < instUse.FirstElement+instUse.NumElements; ++i)
+          localAvail.set(i, DIKind::Yes);
+      }
+
+      // Have we found the instruction that initializes TheMemory?
+      if (localAvail.isAllYes()) {
+        BBI++; // bump to next instruction for an insertion point.
+        break;
+      }
+    }
+
+    // Make sure we found the initializing use of TheMemory.
+    if (BBI == block->end())
+      llvm::report_fatal_error("this block is not initializing?");
+
+    // TODO: obtain `self` from the entry-block and insert the hop.
+    // probably need to emit a borrow and then clean up injectHopAfterBuiltinInit, etc.
+    llvm::dbgs() << "would insert hop_to_executor before: ";
```

Comment:
All of these dbgs things need to be guarded with LLVM_DEBUG.

Reply:
OK. that line will be gone soon though.